migrations/migration.go

```diff
@@ -387,13 +390,48 @@ func (m *StorageMigration) MigrateNestedValue(
 				inter.RemoveReferencedSlab(existingValueStorable)
 			}
 
-			if dictionary.ContainsKey(inter, emptyLocationRange, keyToSet) {
-				panic(errors.NewUnexpectedError(
-					"dictionary contains new key after removal of old key (conflict): %s",
+			// Before inserting the new key-value pair, check if the key already exists.
+			// If it still exists, even after we removed the old key,
+			// then there's a conflict, before the migration, some value was already stored with the new key.
+			// Move the key-value pair into a new dictionary under a new unique storage path, and report it
+
+			existingKeyStorable, _ = dictionary.RemoveWithoutTransfer(
+				inter,
+				emptyLocationRange,
+				keyToSet,
+			)
+			if existingKeyStorable != nil {
+				owner := dictionary.GetOwner()
+
+				storageMap := m.storage.GetStorageMap(owner, common.PathDomainStorage.Identifier(), true)
+				conflictDictionary := interpreter.NewDictionaryValueWithAddress(
+					inter,
+					emptyLocationRange,
+					dictionary.Type,
+					owner,
+				)
+				conflictDictionary.InsertWithoutTransfer(
+					inter,
+					emptyLocationRange,
 					keyToSet,
-				))
+					// NOTE: move the *existing* value to the conflict dictionary,
+					// NOT the value that was just removed – the removed value is the value to set
+					existingValue,
+				)
+
+				conflictStorageMapKey := m.nextDictionaryKeyConflictStorageMapKey()
+
+				storageMap.SetValue(
+					inter,
+					conflictStorageMapKey,
+					conflictDictionary,
+				)
+
+				reporter.DictionaryKeyConflict(conflictStorageMapKey)
 			}
 
+			// Finally, insert the new key-value pair
+
```

Comment:
But we always want to insert the migrated value, no?

The idea is that if the input state contains a dictionary with
- `key1: value1` 
- `key2: value2`

then `key1` is migrated to `key1_migrated`, and `value1` is migrated to `value1_migrated`. 

If `key1_migrated` happens to be equal to `key2`, then we have a conflict – we should not overwrite `key2`/`value2` when inserting `key1_migrated`/`value1_migrated`.

So we remove `key2`/`value2` and move it to a new dictionary.

I think the part that is currently missing is that `key2`/`value2` also need to be migrated, even if gets moved.

Reply:
> I think the part that is currently missing is that key2/value2 also need to be migrated, even if gets moved.

hmm, the catch is `key2`/`value2` may have already been migrated and inserted back, prior to `key1`/`value1`, depending on the order. It could get more complex if we try to migrate the 'removed' value. I think it could be simpler to move the currently migrating value (value at hand) always.

Reply:
Great idea! I was so focused on the approach "remove key2/value2 and move it to a new dictionary" that I didn't understand your comments at first. Moving the currently migrated key/value is a much better idea 👍 

Will push a change

Reply:
Tried to improve this in 5bd7d47dc7010aee7f2fde83bdcddf3388ddc814. @SupunS could you PTAL?

Reply:
nice, looks good!